File: mobile/ios/Tests/CertificatePinningTests.swift
import XCTest
@testable import ALADDINMobile

class CertificatePinningTests: XCTestCase {
    
    var certificatePinningManager: CertificatePinningManager!
    var networkManager: ALADDINNetworkManager!
    
    override func setUp() {
        super.setUp()
        certificatePinningManager = CertificatePinningManager.shared
        networkManager = ALADDINNetworkManager.shared
    }
    
    override func tearDown() {
        certificatePinningManager = nil
        networkManager = nil
        super.tearDown()
    }
    
    // MARK: - Certificate Pinning Tests
    
    func testCertificatePinningManagerInitialization() {
        // Проверяем, что менеджер инициализируется
        XCTAssertNotNil(certificatePinningManager)
    }
    
    func testPinnedHostsList() {
        // Проверяем, что список защищенных хостов не пустой
        let status = certificatePinningManager.getPinningStatus()
        XCTAssertFalse(status.isEmpty)
        
        // Проверяем наличие основных хостов
        XCTAssertTrue(status.keys.contains("api.aladdin.security"))
        XCTAssertTrue(status.keys.contains("ai.aladdin.security"))
        XCTAssertTrue(status.keys.contains("vpn.aladdin.security"))
    }
    
    func testCertificateValidation() {
        // Создаем мок сертификат для тестирования
        let mockTrust = createMockSecTrust()
        let host = "api.aladdin.security"
        
        // Тестируем валидацию (в реальном тесте здесь был бы настоящий сертификат)
        let isValid = certificatePinningManager.validateCertificate(mockTrust, host: host)
        
        // В тестовой среде ожидаем false, так как используем мок
        XCTAssertFalse(isValid)
    }
    
    func testInvalidHostValidation() {
        let mockTrust = createMockSecTrust()
        let invalidHost = "malicious-site.com"
        
        let isValid = certificatePinningManager.validateCertificate(mockTrust, host: invalidHost)
        XCTAssertFalse(isValid)
    }
    
    // MARK: - Network Manager Tests
    
    func testNetworkManagerInitialization() {
        XCTAssertNotNil(networkManager)
    }
    
    func testSecureRequest() {
        let expectation = XCTestExpectation(description: "Secure request")
        
        // Тестируем безопасный запрос
        networkManager.secureRequest(
            "test",
            method: .get,
            responseType: TestResponse.self
        ) { result in
            switch result {
            case .success:
                XCTFail("Expected failure in test environment")
            case .failure(let error):
                // Ожидаем ошибку в тестовой среде
                XCTAssertNotNil(error)
            }
            expectation.fulfill()
        }
        
        wait(for: [expectation], timeout: 5.0)
    }
    
    func testPinningStatus() {
        let status = networkManager.getPinningStatus()
        XCTAssertFalse(status.isEmpty)
    }
    
    // MARK: - Helper Methods
    
    private func createMockSecTrust() -> SecTrust {
        // Создаем мок SecTrust для тестирования
        var trust: SecTrust?
        let result = SecTrustCreateWithCertificates(
            createMockCertificate(),
            SecPolicyCreateSSL(true, "api.aladdin.security" as CFString),
            &trust
        )
        
        XCTAssertEqual(result, errSecSuccess)
        return trust!
    }
    
    private func createMockCertificate() -> SecCertificate {
        // Создаем мок сертификат для тестирования
        let mockCertData = Data("mock_certificate_data".utf8)
        let cert = SecCertificateCreateWithData(nil, mockCertData as CFData)
        return cert!
    }
}

// MARK: - Test Response Model
struct TestResponse: Codable {
    let message: String
    let success: Bool
}

// MARK: - Performance Tests
class CertificatePinningPerformanceTests: XCTestCase {
    
    func testCertificateValidationPerformance() {
        let manager = CertificatePinningManager.shared
        let mockTrust = createMockSecTrust()
        
        measure {
            for _ in 0..<1000 {
                _ = manager.validateCertificate(mockTrust, host: "api.aladdin.security")
            }
        }
    }
    
    private func createMockSecTrust() -> SecTrust {
        var trust: SecTrust?
        let result = SecTrustCreateWithCertificates(
            createMockCertificate(),
            SecPolicyCreateSSL(true, "api.aladdin.security" as CFString),
            &trust
        )
        
        XCTAssertEqual(result, errSecSuccess)
        return trust!
    }
    
    private func createMockCertificate() -> SecCertificate {
        let mockCertData = Data("mock_certificate_data".utf8)
        let cert = SecCertificateCreateWithData(nil, mockCertData as CFData)
        return cert!
    }
}

File: mobile/android/Tests/CertificatePinningTests.kt
package com.aladdin.mobile.tests

import android.content.Context
import androidx.test.core.app.ApplicationProvider
import com.aladdin.mobile.security.CertificatePinningManager
import com.aladdin.mobile.network.ALADDINNetworkManager
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config
import java.security.cert.X509Certificate
import javax.inject.Inject
import kotlin.test.assertNotNull
import kotlin.test.assertTrue
import kotlin.test.assertFalse

@RunWith(RobolectricTestRunner::class)
@Config(sdk = [28])
class CertificatePinningTests {
    
    @Inject
    lateinit var certificatePinningManager: CertificatePinningManager
    
    @Inject
    lateinit var networkManager: ALADDINNetworkManager
    
    private lateinit var context: Context
    
    @Before
    fun setUp() {
        context = ApplicationProvider.getApplicationContext()
        
        // В реальном тесте здесь была бы инициализация Dagger
        // certificatePinningManager = CertificatePinningManager(context)
        // networkManager = ALADDINNetworkManager(context, certificatePinningManager)
    }
    
    // MARK: - Certificate Pinning Tests
    
    @Test
    fun testCertificatePinningManagerInitialization() {
        // Проверяем, что менеджер инициализируется
        assertNotNull(certificatePinningManager)
    }
    
    @Test
    fun testPinnedHostsList() {
        // Проверяем, что список защищенных хостов не пустой
        val status = certificatePinningManager.getPinningStatus()
        assertTrue(status.isNotEmpty())
        
        // Проверяем наличие основных хостов
        assertTrue(status.containsKey("api.aladdin.security"))
        assertTrue(status.containsKey("ai.aladdin.security"))
        assertTrue(status.containsKey("vpn.aladdin.security"))
    }
    
    @Test
    fun testCertificateValidation() {
        // Создаем мок сертификат для тестирования
        val mockCertificate = createMockCertificate()
        val host = "api.aladdin.security"
        
        // Тестируем валидацию (в реальном тесте здесь был бы настоящий сертификат)
        val isValid = certificatePinningManager.validateCertificate(host, mockCertificate)
        
        // В тестовой среде ожидаем false, так как используем мок
        assertFalse(isValid)
    }
    
    @Test
    fun testInvalidHostValidation() {
        val mockCertificate = createMockCertificate()
        val invalidHost = "malicious-site.com"
        
        val isValid = certificatePinningManager.validateCertificate(invalidHost, mockCertificate)
        assertFalse(isValid)
    }
    
    // MARK: - Network Manager Tests
    
    @Test
    fun testNetworkManagerInitialization() {
        assertNotNull(networkManager)
    }
    
    @Test
    fun testSecureRequest() {
        // Тестируем безопасный запрос
        val result = runBlocking {
            networkManager.secureRequest(
                endpoint = "test",
                method = com.aladdin.mobile.network.HttpMethod.GET,
                responseType = TestResponse::class.java
            )
        }
        
        // Ожидаем ошибку в тестовой среде
        assertTrue(result.isFailure)
    }
    
    @Test
    fun testPinningStatus() {
        val status = networkManager.getPinningStatus()
        assertTrue(status.isNotEmpty())
    }
    
    @Test
    fun testCertificatePinnerCreation() {
        val secureClient = certificatePinningManager.createSecureClient()
        assertNotNull(secureClient.certificatePinner)
    }
    
    // MARK: - Performance Tests
    
    @Test
    fun testCertificateValidationPerformance() {
        val manager = certificatePinningManager
        val mockCertificate = createMockCertificate()
        
        val startTime = System.currentTimeMillis()
        
        repeat(1000) {
            manager.validateCertificate("api.aladdin.security", mockCertificate)
        }
        
        val endTime = System.currentTimeMillis()
        val duration = endTime - startTime
        
        // Проверяем, что валидация выполняется быстро (менее 1 секунды для 1000 операций)
        assertTrue(duration < 1000, "Certificate validation too slow: ${duration}ms")
    }
    
    // MARK: - Helper Methods
    
    private fun createMockCertificate(): X509Certificate {
        // Создаем мок сертификат для тестирования
        val certFactory = java.security.cert.CertificateFactory.getInstance("X.509")
        val certData = "mock_certificate_data".toByteArray()
        val inputStream = java.io.ByteArrayInputStream(certData)
        
        return certFactory.generateCertificate(inputStream) as X509Certificate
    }
    
    private suspend fun <T> runBlocking(block: suspend () -> T): T {
        return block()
    }
}

// MARK: - Test Response Model
data class TestResponse(
    val message: String,
    val success: Boolean
)

// MARK: - Integration Tests
class CertificatePinningIntegrationTests {
    
    @Test
    fun testEndToEndCertificatePinning() {
        // Интеграционный тест для проверки всего процесса pinning
        val context = ApplicationProvider.getApplicationContext()
        val manager = CertificatePinningManager(context)
        
        // Проверяем инициализацию
        assertNotNull(manager)
        
        // Проверяем создание клиента
        val client = manager.createSecureClient()
        assertNotNull(client)
        
        // Проверяем статус pinning
        val status = manager.getPinningStatus()
        assertTrue(status.isNotEmpty())
        
        // Проверяем обновление сертификатов
        manager.updateCertificates() // Не должно вызывать исключений
    }
    
    @Test
    fun testNetworkManagerWithPinning() {
        val context = ApplicationProvider.getApplicationContext()
        val pinningManager = CertificatePinningManager(context)
        val networkManager = ALADDINNetworkManager(context, pinningManager)
        
        // Проверяем инициализацию
        assertNotNull(networkManager)
        
        // Проверяем статус pinning
        val status = networkManager.getPinningStatus()
        assertTrue(status.isNotEmpty())
        
        // Проверяем обновление сертификатов
        networkManager.updateCertificates() // Не должно вызывать исключений
    }
}

